arroyo/processing/strategies/routing_producer.py

```diff
@@ -0,0 +1,156 @@
+    3. A message router that routes messages to a topic based on which slice
+    the message needs to be sent to.
+    """
+
+    @abstractmethod
+    def get_route_for_message(self, message: Message[KafkaPayload]) -> MessageRoute:
+        """
+        This method must return the MessageRoute on which the message should
+        be produced. Implementations of this method can vary based on the
+        specific use case.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def shutdown(self, timeout: Optional[float] = None) -> None:
+        """
+        Shutdown the router. This method is called when join() is called on
+        the strategy. This method can be used to perform any cleanup needed
+        by the router. Most often that would always include closing the
+        producers that are used by the router.
+        """
+        raise NotImplementedError
+
+
+class RoutingProducerStep(ProcessingStrategy[KafkaPayload]):
+    """
+    This strategy is used to route messages to different producers/topics
+    based on the message payload. It delegates the routing logic to the
+    message router class which is passed in as a parameter. Please look at
+    the documentation for the MessageRouter class for more details.
+
+    The strategy also does not own any producers. The producers are owned by
+    the message router. As such, the strategy does not close the producers.
+    So it is important that the message router closes the producers when its
+    shutdown method is called. shutdown() is not called during the close()
+    method as it has to be non-blocking. The shutdown() method is therefore
+    called during the join() method.
+    """
+
+    def __init__(
+        self,
+        commit_function: Commit,
+        message_router: MessageRouter,
+    ) -> None:
+        self.__commit_function = commit_function
+        self.__message_router = message_router
+        self.__closed = False
+        self.__offsets_to_be_committed: MutableMapping[Partition, Position] = {}
+        self.__queue: Deque[
+            Tuple[Message[KafkaPayload], Future[Message[KafkaPayload]]]
+        ] = deque()
+
+    def poll(self) -> None:
+        """
+        Periodically check which messages have been produced successfully and
+        call the commit function for the offsets that have been processed.
+        In order to commit offsets we need strict ordering, hence we stop
+        processing the queue of pending futures as soon as we encounter one
+        which is not yet completed.
+        """
+        while self.__queue:
+            message, future = self.__queue[0]
+
+            if not future.done():
+                break
+
+            exc = future.exception()
+            if exc is not None:
+                raise exc
+
+            self.__queue.popleft()
+            self.__offsets_to_be_committed[message.partition] = Position(
+                message.next_offset, message.timestamp
+            )
```

Comment:
Please just call `self.__commit_function({message.partition: message.position_to_commit})` here and remove `self.__offsets_to_be_committed` entirely